Static cctor analysis for reflection

vitek-karas opened the pull request:
This brings the NativeAOT behavior around static cctor analysis on part with illink.

Main change is to add a new node which checks for RUC/RAF/RDC on static cctor and produces a warning if there is one (as these are not allowed on static cctor). The node is added for every static cctor method which has its entry point included in the app.

The rest of the changes are just minor fixes and adjusting tests to the new behavior.

Contributes to https://github.com/dotnet/runtime/issues/82447.

vitek-karas committed:
Implements correct data flow analysis around static cctors

The main open question of this change is where to trigger the data flow for static cctors from.

EEType presence is not enough, if the only thing access on the type is a field there doesn't seem to be EEType generated for it. So I added a hook from NonGCStatics. But it seems that a better solution would be to introduce a notion of ".cctor access" into the metadata manager and call it from all the places which actually get a cctor entry point (3 places).

vitek-karas committed:
Introduce a new node for static ctor analysis and move away from eetypes

vitek-karas committed:
Revert unnecessary change

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs:
```diff
@@ -563,6 +563,8 @@ protected override void GetDependenciesDueToMethodCodePresenceInternal(ref Depen
                     AddDataflowDependency(ref dependencies, factory, methodIL, "Method has annotated parameters");
                 }
+
+                StaticConstructorAnalysisNode.GetDependencies(ref dependencies, factory, method);
```
Why don't we do the check here? (Why is the new node needed?)

vitek-karas replied:
Can we guarantee that this is only called once per method? The new node is to avoid duplicate warnings.

MichalStrehovsky replied:
I think so - this callback is called when a method body is generated. We only generate them once.

vitek-karas replied:
OK- I'll look around and make the change then.

vitek-karas replied:
It can happen that it will be called more than once (I think) for GVMs, the `RuntimeMethodHandleNode` will call this on the cannon abstract GVM method - and that can probably happen more than once if it finds more than one instantiation.

But since the new code only cares about static cctor, that has no effect (cctor is never going to be a GVM).

I simplified this change by removing the new node.

vitek-karas committed:
Remove the new node, it's likely not needed.

The only case where it might matter is if there's a handle access to a GVM. Then we go through this code:
https://github.com/dotnet/runtime/blob/57ae91cf6fc5672e34705b1a272cf268761d505a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/RuntimeMethodHandleNode.cs#L60

That will potentially call the `GetDependenciesDueToMethodCodePresence` multiple times on the same method.

But that should not affect static cctors (since they're not GVMs) so the `GetDependenciesDueToMethodCodePresense` should only be called once per static cctor.

And even if it does happen (in the future) all it would cause is potentially generating duplicate warnings, no real functional problems.

vitek-karas commented:
@MichalStrehovsky could you please take another look - I think this is ready.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs:
```diff
@@ -198,6 +198,9 @@ internal void MarkStaticConstructor(in MessageOrigin origin, TypeDesc type, stri
                 return;
 
+            if (type.HasStaticConstructor)
+                CheckAndWarnOnReflectionAccess(origin, type.GetStaticConstructor());
```
Will this cause two warnings to be generated? If I understand this correctly, this will generate a warning for reflection access to cctor, and then we get another one when we generate the method body, because this attribute on a cctor is always an error.

vitek-karas replied:
It actually won't - we intentionally ignore the `Requires*` attributes on static cctor for generating warnings (basically relying on the fact that they will produce a warning just due to their presence, without actual callsite).
https://github.com/dotnet/runtime/blob/24fa97ab99b4a63f638f63a90ae83c4b80772114/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs#L100
https://github.com/dotnet/runtime/blob/24fa97ab99b4a63f638f63a90ae83c4b80772114/src/tools/illink/src/linker/Linker/Annotations.cs#L609
https://github.com/dotnet/runtime/blob/24fa97ab99b4a63f638f63a90ae83c4b80772114/src/tools/illink/src/ILLink.RoslynAnalyzer/RequiresISymbolExtensions.cs#L60

vitek-karas replied:
I'm adding a test for this behavior as well.

vitek-karas committed:
Add test for accessing RUC cctor directly
